### ISSUE TITLE
bad check_types two objects

### ISSUE THREAD

**Issue by paulsimon** [google/closure-compiler#961]
I am sorry about creating new issue.

```javascript
/**
 * @typedef {{ value: Array<string> }}
 */
var firstType;

/**
 * @typedef { string }
 */
var good;

/**
 * @typedef {{ value: string }}
 */
var bad;

/**
 * @typedef { Array<(firstType|bad)> }
 */
var list;

/**
 * @param {(firstType|bad)} x
 */
var addTo = function(x) {
  list.push(x);
};

var item = { value: ['33'] };
addTo(item);
```

I use object and string (firstType and good) check_types is ok. Vide online compiler: http://goo.gl/rebsT9
I use two objects (firstType and bad) check_types return warning. Vide online: http://goo.gl/h80N9u

I can for good check_types use @type, however it is falsity and check_types is functionless.
```javascript
var item = /** @type {bad} */ ({ value: ['33'] });
addTo(item);
```

**Comment by dimvar** [comment]
Yeah this seems to be a bug in the current type inference. If you explicitly annotate the type of item as <code> /** @type {firstType} */ </code> then it works.

We're not prioritizing bugs in the current type checker at the moment, so this will probably not be fixed soon.

**Comment by concavelenz** [comment]
What is happening here is that "item" has a inferred record type.  The type starts as "{value: !Array<string>}".  When type inference tries to match 
"{value: Array<string>}" it expands the "value" property type to include "null" (as record types are invariant it wouldn't match otherwise).  So far all is good but there is another record type in the union: "{value:string}".  When it tries to match this it again tries to expand the field definition to try to create a match so now you have ""{value: (Array<string>|string)}".
